File: python/zebra-puzzle/zebra_puzzle.py
"""
Algorithm
---------

"Guided search" or "mutating search" based on kahgoh's Elixir solution
as described by Jeremy and Erik.

This algorithm starts with a random street as the search space.
It checks that all the rules match.
If a rule does not match, it modifies the house(s) that violates the rules
so that the rule would pass and adds the new street to the search space.
Repeat until a valid street is found.

For example, consider "The Englishman lives in the red house."
Find the house where the Englishman lives and find the red house.
If they are the same house, the rule passes and we check the next rule.
If they are different, try either swapping the nationalities between those house
or swapping the colors.
After either of those two swaps, this rule should now pass.

Representation
--------------
A street/combo is a tuple of five houses.
A house is a collection of five attributes.
An attribute is an integer [0..5).
A house combines attributes via bit shifting, allocating 3 bits per attribute.

Rules
-----

1. There are five houses.
2. The Englishman lives in the red house.
3. The Spaniard owns the dog.
4. Coffee is drunk in the green house.
5. The Ukrainian drinks tea.
6. The green house is immediately to the right of the ivory house.
7. The Old Gold smoker owns snails.
8. Kools are smoked in the yellow house.
9. Milk is drunk in the middle house.
10. The Norwegian lives in the first house.
11. The man who smokes Chesterfields lives in the house next to the man with the fox.
12. Kools are smoked in the house next to the house where the horse is kept.
13. The Lucky Strike smoker drinks orange juice.
14. The Japanese smokes Parliaments.
15. The Norwegian lives next to the blue house.

Each of the five houses is painted a different color, and their
inhabitants are of different national extractions, own different pets,
drink different beverages and smoke different brands of cigarettes.
"""

import random

# How far to bit shift each attribute, number of bits.
NATIONALITY, PET, COLOR, DRINK, SMOKE = [i * 3 for i in range(5)]
NATIONALITIES = ENGLISHMAN, SPANIARD, JAPANESE, UKRAINIAN, NORWEGIAN = [i << NATIONALITY for i in range(5)]
PETS = DOG, FOX, SNAILS, HORSE, ZEBRA = [i << PET for i in range(5)]
COLORS = RED, IVORY, GREEN, BLUE, YELLOW = [i << COLOR for i in range(5)]
DRINKS = TEA, COFFEE, ORANGE_JUICE, MILK, WATER = [i << DRINK for i in range(5)]
SMOKES = OLD_GOLD, KOOLS, LUCKY_STRIKE, PARLIAMENTS, CHESTERFIELDS = [i << SMOKE for i in range(5)]

RuleResult = tuple[bool, list[tuple[int, ...]]]


def matches(house: int, attribute: int, value: int) -> bool:
    """Return if a house has a specific attribute, eg NATIONALITY == NORWEGIAN."""
    return house & (0b111 << attribute) == value


def find_house(combo: tuple[int, ...], attribute: int, value: int) -> int:
    """Return the index of a house having a specific value for an attribute."""
    mask = 0b111 << attribute
    return next(idx for idx, house in enumerate(combo) if (house & mask) == value)


def swap(combo: tuple[int, ...], a, b, attribute) -> tuple[int, ...]:
    """Swap attributes between two houses, eg swap the NATIONALITY of houses 1 and 2."""
    # mask is used to get a specific attribute. rev_mask is used to zero out the attribute.
    mask = 0b111 << attribute
    rev_mask  = (2 << 16) - 1 - mask
    # Update the houses.
    houses = list(combo)
    c = houses[a]
    # a becomes a with one attribute taken from b.
    # b becomes b with one attribute taken from c.
    houses[a] = (houses[a] & rev_mask) | houses[b] & mask
    houses[b] = (houses[b] & rev_mask) | c & mask
    return tuple(houses)


def house_rule(combo: tuple[int, ...], attribute_a: int, val_a: int, attribute_b: int, val_b: int,) -> RuleResult:
    """Check a single house has two specific attributes."""
    a = find_house(combo, attribute_a, val_a)
    b = find_house(combo, attribute_b, val_b)
    if a == b:
        return True, []
    return False, [swap(combo, a, b, attribute_a), swap(combo, a, b, attribute_b)]


def adjacent_rule(combo: tuple[int, ...], attribute_a: int, val_a: int, attribute_b: int, val_b: int) -> RuleResult:
    """Check adjacent houses have specific attributes."""
    a = find_house(combo, attribute_a, val_a)
    b = find_house(combo, attribute_b, val_b)
    # Adjacent
    if abs(a - b) == 1:
        return True, []
    alternatives = []
    # Swap attributes with adjacent houses. Two to four possibilities.
    if a > 0:
        alternatives.append(swap(combo, a - 1, b, attribute_b))
    if a < 4:
        alternatives.append(swap(combo, a + 1, b, attribute_b))
    if b > 0:
        alternatives.append(swap(combo, b - 1, b, attribute_a))
    if b < 4:
        alternatives.append(swap(combo, b + 1, b, attribute_a))
    return False, alternatives


def green_ivory(combo: tuple[int, ...]) -> RuleResult:
    """6. The green house is immediately to the right of the ivory house."""
    a = find_house(combo, COLOR, GREEN)
    b = find_house(combo, COLOR, IVORY)
    if a == b + 1:
        return True, []
    alternatives = []
    if a > 0:
        alternatives.append(swap(combo, a - 1, b, COLOR))
    if b < 4:
        alternatives.append(swap(combo, b + 1, a, COLOR))
    return False, alternatives


def milk_middle(combo: tuple[int, ...]) -> RuleResult:
    """9. Milk is drunk in the middle house."""
    a = 2  # middle
    b = find_house(combo, DRINK, MILK)
    if a == b:
        return True, []
    return False, [swap(combo, a, b, DRINK)]


def norway_first(combo: tuple[int, ...]) -> RuleResult:
    """10. The Norwegian lives in the first house."""
    a = 0  # first
    b = find_house(combo, NATIONALITY, NORWEGIAN)
    if a == b:
        return True, []
    return False, [swap(combo, a, b, NATIONALITY)]


RULES = [
    # 2. The Englishman lives in the red house.
    lambda x: house_rule(x, NATIONALITY, ENGLISHMAN, COLOR, RED),
    # 3. The Spaniard owns the dog.
    lambda x: house_rule(x, NATIONALITY, SPANIARD, PET, DOG),
    # 4. Coffee is drunk in the green house.
    lambda x: house_rule(x, DRINK, COFFEE, COLOR, GREEN),
    # 5. The Ukrainian drinks tea.
    lambda x: house_rule(x, NATIONALITY, UKRAINIAN, DRINK, TEA),
    # 7. The Old Gold smoker owns snails.
    lambda x: house_rule(x, SMOKE, OLD_GOLD, PET, SNAILS),
    # 8. Kools are smoked in the yellow house.
    lambda x: house_rule(x, SMOKE, KOOLS, COLOR, YELLOW),
    # 13. The Lucky Strike smoker drinks orange juice.
    lambda x: house_rule(x, SMOKE, LUCKY_STRIKE, DRINK, ORANGE_JUICE),
    # 14. The Japanese smokes Parliaments.
    lambda x: house_rule(x, NATIONALITY, JAPANESE, SMOKE, PARLIAMENTS),
    # 6. The green house is immediately to the right of the ivory house.
    green_ivory,
    # 9. Milk is drunk in the middle house.
    milk_middle,
    # 10. The Norwegian lives in the first house.
    norway_first,
    # 11. The man who smokes Chesterfields lives in the house next to the man with the fox.
    lambda x: adjacent_rule(x, SMOKE, CHESTERFIELDS, PET, FOX),
    # 12. Kools are smoked in the house next to the house where the horse is kept.
    lambda x: adjacent_rule(x, SMOKE, KOOLS, PET, HORSE),
    # 15. The Norwegian lives next to the blue house.
    lambda x: adjacent_rule(x, NATIONALITY, NORWEGIAN, COLOR, BLUE),
]


def solve_puzzle() -> tuple[int, ...]:
    """Solve the puzzle."""
    # Create a random street.
    attributes = [NATIONALITIES.copy(), PETS.copy(), COLORS.copy(), DRINKS.copy(), SMOKES.copy()]
    for attrib in attributes:
        random.shuffle(attrib)
    houses = [sum(attribs) for attribs in zip(*attributes)]

    # Depth first. If a street doesn't fulfill a role, try swapping in either direction to match the rule and queue both.
    combos = [tuple(houses)]
    seen = set(combos)
    while combos:
        combo = combos.pop()
        for rule in RULES:
            ok, alternatives = rule(combo)
            if not ok:
                for alternative in alternatives:
                    if alternative in seen:
                        continue
                    seen.add(alternative)
                    combos.append(alternative)
                break
        else:
            # All the rules pass.
            return combo
    raise RuntimeError("not solved")


def nationality(house: int) -> str:
    n = house & (0b111 << NATIONALITY)
    return {
        ENGLISHMAN: "Englishman",
        SPANIARD: "Spaniard",
        JAPANESE: "Japanese",
        UKRAINIAN: "Ukrainian",
        NORWEGIAN: "Norwegian",
    }[n]


def drinks_water() -> str:
    """Return the nationality of the water drinker."""
    street = solve_puzzle()
    return nationality(street[find_house(street, DRINK, WATER)])


def owns_zebra() -> str:
    """Return the nationality of the zebra owner."""
    street = solve_puzzle()
    return nationality(street[find_house(street, PET, ZEBRA)])

File: go/zebra-puzzle/zebra_puzzle.go
// zebra solution based on kahgoh's solution in Elixir.
package zebra

import rand "math/rand/v2"

const (
	ShiftNationality = 0
	ShiftPet         = 3
	ShiftColor       = 6
	ShiftDrink       = 9
	ShiftSmoke       = 12
	MaskNationality  = 0b111 << ShiftNationality
	MaskPet          = 0b111 << ShiftPet
	MaskColor        = 0b111 << ShiftColor
	MaskDrink        = 0b111 << ShiftDrink
	MaskSmoke        = 0b111 << ShiftSmoke
	MaskAll          = MaskNationality | MaskPet | MaskColor | MaskDrink | MaskSmoke
	Englishman       = 1 << ShiftNationality
	Spaniard         = 2 << ShiftNationality
	Japanese         = 3 << ShiftNationality
	Ukrainian        = 4 << ShiftNationality
	Norwegian        = 5 << ShiftNationality
	Dog              = 1 << ShiftPet
	Fox              = 2 << ShiftPet
	Snails           = 3 << ShiftPet
	Horse            = 4 << ShiftPet
	Zebra            = 5 << ShiftPet
	Red              = 1 << ShiftColor
	Ivory            = 2 << ShiftColor
	Green            = 3 << ShiftColor
	Blue             = 4 << ShiftColor
	Yellow           = 5 << ShiftColor
	Tea              = 1 << ShiftDrink
	Coffee           = 2 << ShiftDrink
	OrangeJuice      = 3 << ShiftDrink
	Milk             = 4 << ShiftDrink
	Water            = 5 << ShiftDrink
	OldGold          = 1 << ShiftSmoke
	Kools            = 2 << ShiftSmoke
	LuckyStrike      = 3 << ShiftSmoke
	Parliaments      = 4 << ShiftSmoke
	Chesterfields    = 5 << ShiftSmoke
)

// Street has 5 houses, each a uint16.
type Street [5]uint16

func abs(a int) int {
	if a >= 0 {
		return a
	}
	return -a
}

func nationality(house uint16) string {
	nations := map[uint16]string{
		Englishman: "Englishman",
		Spaniard:   "Spaniard",
		Japanese:   "Japanese",
		Ukrainian:  "Ukrainian",
		Norwegian:  "Norwegian",
	}
	for k, v := range nations {
		if house&MaskNationality == k {
			return v
		}
	}
	return ""
}

func findHouse(street Street, mask, value uint16) int {
	for idx, house := range street {
		if house&mask == value {
			return idx
		}
	}
	return -1
}

func swap(street Street, idxA, idxB int, mask uint16) Street {
	revMask := MaskAll ^ mask
	copy := street
	copy[idxA] = street[idxA]&revMask | street[idxB]&mask
	copy[idxB] = street[idxB]&revMask | street[idxA]&mask
	return copy
}

func houseRule(street Street, maskA, valA, maskB, valB uint16) ([]Street, bool) {
	a := findHouse(street, maskA, valA)
	b := findHouse(street, maskB, valB)
	if a == b {
		return nil, true
	}
	return []Street{swap(street, a, b, maskA), swap(street, a, b, maskB)}, false
}

func neighborRule(street Street, maskA, valA, maskB, valB uint16) ([]Street, bool) {
	a := findHouse(street, maskA, valA)
	b := findHouse(street, maskB, valB)
	if abs(a-b) == 1 {
		return nil, true
	}
	alt := []Street{}
	if a > 0 {
		alt = append(alt, swap(street, a-1, b, maskB))
	}
	if a < 4 {
		alt = append(alt, swap(street, a+1, b, maskB))
	}
	if b > 0 {
		alt = append(alt, swap(street, b-1, a, maskA))
	}
	if b < 4 {
		alt = append(alt, swap(street, b+1, a, maskA))
	}
	return alt, false
}

func ruleGreenIvory(street Street) ([]Street, bool) {
	a := findHouse(street, MaskColor, Green)
	b := findHouse(street, MaskColor, Ivory)
	if a == b+1 {
		return nil, true
	}
	alt := []Street{}
	if a > 0 {
		alt = append(alt, swap(street, a-1, b, MaskColor))
	}
	if b < 4 {
		alt = append(alt, swap(street, b+1, a, MaskColor))
	}
	return alt, false
}

func ruleMilkMiddle(street Street) ([]Street, bool) {
	a := 2
	b := findHouse(street, MaskDrink, Milk)
	if a == b {
		return nil, true
	}
	return []Street{swap(street, a, b, MaskDrink)}, false
}

func ruleNorwayFirst(street Street) ([]Street, bool) {
	a := 0
	b := findHouse(street, MaskNationality, Norwegian)
	if a == b {
		return nil, true
	}
	return []Street{swap(street, a, b, MaskNationality)}, false
}

var rules = []func(Street) ([]Street, bool){
	func(s Street) ([]Street, bool) { return houseRule(s, MaskNationality, Englishman, MaskColor, Red) },
	func(s Street) ([]Street, bool) { return houseRule(s, MaskNationality, Spaniard, MaskPet, Dog) },
	func(s Street) ([]Street, bool) { return houseRule(s, MaskDrink, Coffee, MaskColor, Green) },
	func(s Street) ([]Street, bool) { return houseRule(s, MaskNationality, Ukrainian, MaskDrink, Tea) },
	func(s Street) ([]Street, bool) { return houseRule(s, MaskSmoke, OldGold, MaskPet, Snails) },
	func(s Street) ([]Street, bool) { return houseRule(s, MaskSmoke, Kools, MaskColor, Yellow) },
	func(s Street) ([]Street, bool) { return houseRule(s, MaskSmoke, LuckyStrike, MaskDrink, OrangeJuice) },
	func(s Street) ([]Street, bool) {
		return houseRule(s, MaskNationality, Japanese, MaskSmoke, Parliaments)
	},
	ruleGreenIvory,
	ruleMilkMiddle,
	ruleNorwayFirst,
	func(s Street) ([]Street, bool) { return neighborRule(s, MaskSmoke, Chesterfields, MaskPet, Fox) },
	func(s Street) ([]Street, bool) { return neighborRule(s, MaskSmoke, Kools, MaskPet, Horse) },
	func(s Street) ([]Street, bool) { return neighborRule(s, MaskNationality, Norwegian, MaskColor, Blue) },
}

func solver() Street {
	attribs := [][]uint16{
		{Englishman, Spaniard, Japanese, Ukrainian, Norwegian},
		{Dog, Fox, Snails, Horse, Zebra},
		{Red, Ivory, Green, Blue, Yellow},
		{Tea, Coffee, OrangeJuice, Milk, Water},
		{OldGold, Kools, LuckyStrike, Parliaments, Chesterfields},
	}
	for _, slice := range attribs {
		rand.Shuffle(5, func(i, j int) { slice[i], slice[j] = slice[j], slice[i] })
	}
	street := Street{}
	for i := range 5 {
		for _, attrib := range attribs {
			street[i] += attrib[i]
		}
	}

	todo := []Street{street}
	seen := map[Street]bool{}
	for len(todo) != 0 {
		street := todo[len(todo)-1]
		todo = todo[:len(todo)-1]
		found := true
		for _, rule := range rules {
			if alternatives, ok := rule(street); !ok {
				for _, alternative := range alternatives {
					if seen[alternative] {
						continue
					}
					seen[alternative] = true
					todo = append(todo, alternative)
				}
				found = false
				break
			}
		}
		if found {
			return street
		}
	}
	return Street{}
}

type Solution struct {
	DrinksWater string
	OwnsZebra   string
}

func SolvePuzzle() Solution {
	street := solver()
	solution := Solution{}
	for _, house := range street {
		if house&MaskDrink == Water {
			solution.DrinksWater = nationality(house)
		}
		if house&MaskPet == Zebra {
			solution.OwnsZebra = nationality(house)
		}
	}
	return solution
}
